### ISSUE TITLE
App Roles Not Available in API Protected by Microsoft Identity

### ISSUE THREAD

**Issue by DigitalPZ** [Azure-Samples/ms-identity-blazor-server#5]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is for a: (mark with an `x`)
```
- [ x] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Minimal steps to reproduce
> Setup a Blazor Server app as a client and an ASP.NET Core Web API and protect them using Microsot Identity framework as per examples here. Insert the following line of code in ConfigureServices in Startup.cs also like the example provided in this repo:
`            JwtSecurityTokenHandler.DefaultMapInboundClaims = false;
`
In Azure AD admin, add 2 new app roles under the app registration for the client. Go to Enterprise applications and select the client app. Select Users and Groups and add an assignment to each app role using the same user account. Run the application and login using the user account that the app roles were assigned to.

When you inspect the user claims in the client you will indeed see a claim for each role assignment created earlier. The type will be "roles" and the value will be whatever you used when you created the app role.

If you inspect the user claims inside the API being called, the "roles" claims are not present and these claims are not present in the token within either app. I even tried adding the same line of code as above to startup.cs and still no roles claims.

### Any log messages given by the failure
> No errors present

### Expected/desired behavior
> I expected to be able to retrieve the roles claims from inside the API code so I can deny access to API's based on app role assignment.

### OS and Version?
> Windows 7, 8 or 10. Linux (which distribution). macOS (Yosemite? El Capitan? Sierra?)

### Versions
>

### Mention any other details that might be useful

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

**Comment by Shama-K** [comment]
If you need to retrieve role claim for API, then it should be added into app registration of API as well.

**Comment by DigitalPZ** [comment]
Shama, does that mean I have to add the app roles to both the Blazor client and the API and then I have to assign users to app roles in both apps? With ASP.NET Core Identity we used to be able to just add claims to the user and they would get coded into the token and be available in both the client and the API. Is there a different way of doing that with the Microsoft Identity framework?

**Comment by Shama-K** [comment]
Can you share the code and details how you handled it?
To be more clear about roles claim from Azure AD, you need to add role claims to the application where you need it. If it's only web API then it's not required in client app.

**Comment by DigitalPZ** [comment]
I haven't been able to locate that old code. I have verified that the token received in the API does contain the roles claims that were assigned to the client app. Those claims however, do not show up in the claims for the ClaimPrincipal that I can retrieve via the HttpContext. Is there a way to use the claims embedded in the token instead of having to add them to the API as well? Ideally we want to define the app roles and assign them only once but be able to access them in both the client app and the API.

**Comment by Shama-K** [comment]
If you define app roles and assign them to users  in client and API registrations on Azure portal, then roles claim will be present in both ID Token and Access Token. You don't need to embed roles claim with code. You can follow [angular-spa-dotnetcore-webapi-roles](https://github.com/Azure-Samples/ms-identity-javascript-angular-spa-dotnetcore-webapi-roles-groups/tree/master/chapter1) sample to get an idea of how it works. For authorization in Blazor, you can follow: [ASP.NET Core Blazor authorization](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/?view=aspnetcore-5.0#authorization). 
Please let me know if it helps to resolve the issue.

**Comment by Shama-K** [comment]
Unfortunately, I'm not aware of this. But for your scenario, if you define app roles and assign them to users  in client and API registrations on Azure portal, then roles claim will be present in both ID Token and Access Token. You don't need to embed roles claim with code. You can follow [angular-spa-dotnetcore-webapi-roles](https://github.com/Azure-Samples/ms-identity-javascript-angular-spa-dotnetcore-webapi-roles-groups/tree/master/chapter1) sample to get an idea of how it works. For authorization in Blazor, you can follow: [ASP.NET Core Blazor authorization](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/?view=aspnetcore-5.0#authorization). 
Please let me know if it helps to resolve the issue.

**Comment by DigitalPZ** [comment]
I now have a follow-up issue. I've added app roles to the API and I've assigned users to various app roles. When the API is called I can inspect the current ClaimPrincipal via HttpContext.User and I can see that there is a claim called roles and it has the name of the role that was assigned. I have tried using an annotation **[Authorize(Roles = "Weather.Read")]** and I have tried using             **var check = user.IsInRole("Weather.Read");** but in both cases false is returned even though the roles claim is clearly there. What am I missing to get this to work?

**Comment by DigitalPZ** [comment]
Never mind, I found it. I was missing the following in Startup.cs in ConfigureServices:
`            services.Configure<JwtBearerOptions>(JwtBearerDefaults.AuthenticationScheme, options =>
            {
                options.TokenValidationParameters.RoleClaimType = "roles";
            });
`
